### PR TITLE
Adds woi support to Alliedvision camera

### PR DESCRIPTION
**Pull request by bodokaiser**
Things to discuss

1. Should we do some checks before we do `self.woi = woi`, e.g.,
- is `woi` a list with four integers?
- is `x + width <=  self.shape[1] and y + height <= self.shape[0]`?
 
2. Typically, `self.woi` equals `self.shape` and we could safe the `if self.woi is not None`?

### CODE REVIEW

**Comment by ichristen** [comment]
Hi @bodokaiser 

I would prefer to do a comprehensive WOI revamp in one go. #33 

For instance, this 'software WOI' should be in the `Camera` superclass such that all cameras without 'hardware WOI' can use it. Also, I thought that `AlliedVision` had hardware WOI support?

And yes, the current standard is a 4-int `(x_base, w, y_base, h)` as you put. I would like to standardize this for all camera using this standard or another sensible option.

**Review by ichristen — CHANGES_REQUESTED** [review]
Some (but not all?) AlliedVision cameras [support hardware WOI (called ROI there)](https://docs.alliedvision.com/Vimba_X/Vimba_X_DeveloperGuide/viewerGuide.html#image-format-tab). This pull request should implement the hardware WOI. Software WOI should be in the `Camera` superclass.

**Comment by bodokaiser** [comment]
> Some (but not all?) AlliedVision cameras [support hardware WOI (called ROI there)](https://docs.alliedvision.com/Vimba_X/Vimba_X_DeveloperGuide/viewerGuide.html#image-format-tab). This pull request should implement the hardware WOI. Software WOI should be in the `Camera` superclass.

I update the PR later for the hardware WOI. Should I open another PR for the Software WOI?

**Comment by bodokaiser** [comment]
`vmbpy` seems to be somewhat picky with the value ranges of `woi` but throws a descriptive error, e.g., `Called 'set()' of Feature 'Height' with invalid value. 266 is not within [8, 248].` or `Called 'set()' of Feature 'OffsetY' with invalid value. 967 is not a multiple of 2, starting at 0`.

Should we just let the errors of `vmbpy` propagate through `slmsuite`?

Also, what should be the behavior if `set_woi(None)`?

**Comment by ichristen** [comment]
Using the old `self.woi` would work if you check for `None`.

Suppose you have a 100x100 camera. I was concerned about a case where you go from x, w, y, h = 0, 100, 0, 100 to 40, 20, 40, 20. If you set x, y first, then the intermediate woi is x, w, y, h = 40, 100, 40, 100 which is out of range and may error as invalid.

**Comment by bodokaiser** [comment]
> Suppose you have a 100x100 camera. I was concerned about a case where you go from x, w, y, h = 0, 100, 0, 100 to 40, 20, 40, 20. If you set x, y first, then the intermediate woi is x, w, y, h = 40, 100, 40, 100 which is out of range and may error as invalid.

Ah, I see, so setting first the height and width then the offset.

In addition, it seems that AlliedVision doesn't allow any integer value for `w,h,x,y` but once certain ranges of multiples of `8`.

File: slmsuite/hardware/cameras/alliedvision.py [hunk 1]
```diff
@@ -293,7 +293,24 @@ def _set_exposure_hw(self, exposure_s):
...
+            self.cam.OffsetX.set(x)
+            self.cam.OffsetY.set(y)
+        except Exception as e:
+            x, w, y, h = self.woi if self.woi is not None else maxwoi
+            self.cam.Height.set(h)
+            self.cam.Width.set(w)
+            self.cam.OffsetX.set(x)
+            self.cam.OffsetY.set(y)
+
+            raise(e)
```
Comment (ichristen) [inline slmsuite/hardware/cameras/alliedvision.py:311]: I don't think this is proper syntax? Maybe just `raise`?

**Review by ichristen — CHANGES_REQUESTED** [review]
I still think you may need to set to a temp small width to avoid corner cases. For instance, with the 100x100 camera from before, the transition from 40, 20, 40, 20 to 0, 100, 0, 100 might fail because the first step sets to 40, 100, 40, 20 which is out of range. Then again, I don't know how AlliedVision handles these errors.

**Comment by bodokaiser** [comment]
I added some test cases via `alliedvision_test.py` to be sure that we cover the edge cases. For the Avlium 1800 U-510m they passed successfully.

**Review by ichristen — APPROVED** [review]
Sorry for my (very) late reply; things have been busy. Looks good! I added a few comments, and am good with pulling.
